perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_userCertDNs.java

```diff
@@ -61,26 +59,38 @@ public Attribute getAttributeValue(PerunSessionImpl sess, User user, AttributeDe
-
-		User user = null;
-		Attribute attrVirtUserCertDNs = null;
-
-		Matcher extSourceTypeX509Matcher = extSourceTypeX509.matcher(message);
-		Matcher addUserExtSourceMatcher = addUserExtSource.matcher(message);
-		Matcher removeUserExtSourceMatcher = removeUserExtSource.matcher(message);
-
-		if(extSourceTypeX509Matcher.find()) {
-			if(addUserExtSourceMatcher.find() || removeUserExtSourceMatcher.find()) {
-				user = perunSession.getPerunBl().getModulesUtilsBl().getUserFromMessage(perunSession, message);
-				if(user != null) {
-					attrVirtUserCertDNs = perunSession.getPerunBl().getAttributesManagerBl().getAttribute(perunSession, user, AttributesManager.NS_USER_ATTR_VIRT + ":userCertDNs");
-					String messageAttributeSet = attrVirtUserCertDNs.serializeToString() + " set for " + user.serializeToString() + ".";
-					resolvingMessages.add(messageAttributeSet);
-				}
-			}
+	public List<AuditEvent> resolveVirtualAttributeValueChange(PerunSessionImpl perunSession, AuditEvent message) throws InternalErrorException, WrongReferenceAttributeValueException, AttributeNotExistsException, WrongAttributeAssignmentException {
+		List<AuditEvent> resolvingMessages = new ArrayList<>();
+		if (message == null) return resolvingMessages;
+
+		if (message instanceof UserExtSourceAddedToUser) {
+			resolvingMessages.addAll(resolveUserExtSourceAddedToUser(perunSession, (UserExtSourceAddedToUser) message));
+		} else if (message instanceof UserExtSourceRemovedFromUser) {
+			resolvingMessages.addAll(resolveUserExtSourceRemovedFromUser(perunSession, (UserExtSourceRemovedFromUser) message));
+		}
+
+		return resolvingMessages;
+	}
+
+	private List<AuditEvent> resolveUserExtSourceAddedToUser(PerunSessionImpl perunSession, UserExtSourceAddedToUser userExtSourceAddedToUser) throws WrongAttributeAssignmentException, InternalErrorException, AttributeNotExistsException {
+		List<AuditEvent> resolvingMessages = new ArrayList<>();
+
+		if (userExtSourceAddedToUser.getUserExtSource().getExtSource() instanceof ExtSourceX509) {
+			User user = userExtSourceAddedToUser.getUser();
+			Attribute attrVirtUserCertDNs = perunSession.getPerunBl().getAttributesManagerBl().getAttribute(perunSession, user, A_U_V_USER_CERT_DNS);
+			resolvingMessages.add(new AttributeSetForUser(attrVirtUserCertDNs, user));
+		}
+
+		return resolvingMessages;
+	}
+
+	private List<AuditEvent> resolveUserExtSourceRemovedFromUser(PerunSessionImpl perunSession, UserExtSourceRemovedFromUser userExtSourceRemovedFromUser) throws WrongAttributeAssignmentException, InternalErrorException, AttributeNotExistsException {
+		List<AuditEvent> resolvingMessages = new ArrayList<>();
+
+		if (userExtSourceRemovedFromUser.getUserExtSource().getExtSource() instanceof ExtSourceX509) {
+			User user = userExtSourceRemovedFromUser.getUser();
+			Attribute attrVirtUserCertDNs = perunSession.getPerunBl().getAttributesManagerBl().getAttribute(perunSession, user, A_U_V_USER_CERT_DNS);
```

Comment:
Same here: `//FIXME: if the attribute value is null or empty, this method should return AttributeRemovedForUser instead`